#  Classification : Nearest Neighbors and Naive Bayes 

## Classification using Nearest Neighbors

(a) Perform k-Nearest neighbours on the given dataset($X_{knn}$ and $y_{knn}$: where $X_{knn}$ stores feature vectors representing the movies and  $y_{knn}$ stores the 0-1 labelling for each movie) for binary classification of movies, for classifiying whether a given movie is a comedy(label 1) or not a comedy(label 0) . Split the dataset into train(80%), validation(10%) and test sets(10%).Run k-Nearest neighbours for different k values (1,3,7,15,31,63). Select the k, using validation set, which returns the best accuracy score. 

(i)  Report all the validation accuracies for all the values of k. 
<br>(ii) Report accuracy score by performing k-NN on the test dataset using the best chosen k value. 

In [201]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")
X=pd.read_csv('/content/drive/My Drive/X/X_knn.csv')
y=pd.read_csv('/content/drive/My Drive/X/y_knn.csv')
X_array=X.to_numpy()

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## KNN from scratch

In [184]:
import statistics
def euclidean_distance(list1, list2):
  list1=list1
  list2=list2
  distance = 0
  for i in range(len(list1)):
    distance += (list1[i] - list2[i])**2
  return np.sqrt(distance)
def prediction(list1,X,y,k):
  distances=[]
  for i in range(len(X)):
    distances.append((i,euclidean_distance(list1,X[i])))
  distances.sort(key=lambda x: x[1],reverse=True)
  predict=[]
  for i in range(k):
    predict.append(y[distances[i][0]][0])
  return statistics.mode(predict)

### The above scratch code takes huge computational time because of which I used inbuilt Knn Algorithm. A sample code is run below to demonstrate the working of above code

In [189]:
a=pd.read_csv('/content/drive/My Drive/X/X_knn.csv',sep=' ',header=None)
a=a.to_numpy()
y=pd.read_csv('/content/drive/My Drive/X/y_knn.csv',sep=' ',header=None)
y=y.to_numpy()

Train = 80%, Test=10% Validation = 10%

In [190]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(a, y, test_size=0.2, random_state=10)
X_test, X_val, y_test, y_val = train_test_split(X_test, y_test, test_size=0.5, random_state=10)

### Predcition of X_val[2] by training using X_train and y_train with k value 3

In [191]:
prediction(X_val[2],X_train,y_train,3)

0.0

In [202]:
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.metrics import accuracy_score
accuracy=[]
for i in [1,3,7,15,31,63]:
  knn = KNeighborsClassifier(n_neighbors=i)
  knn.fit(X_train, y_train) 
  y_pred= knn.predict(X_val)
  accuracy.append((i,accuracy_score(y_pred,y_val)))
accuracy

[(1, 0.836), (3, 0.85), (7, 0.852), (15, 0.847), (31, 0.853), (63, 0.855)]

## Accuracy for k=63 is 0.855 which is maximum

## Hence can be used for predicting the testset.

In [207]:
j=0
for i in [1,3,7,15,31,63]:
  print('Accuracy for k= '+str(i)+' is',accuracy[j][1])
  j=j+1

Accuracy for k= 1 is 0.836
Accuracy for k= 3 is 0.85
Accuracy for k= 7 is 0.852
Accuracy for k= 15 is 0.847
Accuracy for k= 31 is 0.853
Accuracy for k= 63 is 0.855


## Test Dataset

In [208]:
knn = KNeighborsClassifier(n_neighbors=63)
knn.fit(X_train, y_train) 
y_pred= knn.predict(X_test)
accuracy=accuracy_score(y_pred,y_test)
print('Accuracy for test set is ',accuracy)

Accuracy for test set is  0.823


(b) State why using an even value of k in k-NN should not be chosen

### **Answer:** Even k value is not chosen because there is a chance of tie between neighbour classes, then the maximum nearest neighbour class couln't be found. If there are odd number of predicting classes then even k could be used, because tie is not possible in this case. 



## Learning Naive Bayes' classifier  

### From Continuous Distribution of data

Here, the distribution of the data( $X$ represents the datapoints and $Y$ represents the 0-1 binary-class label; where 0 being the negative class and 1 being the positive class) is already known.
<br>Consider the following one-dimensional(1-D) Gaussian distributions where means and variances are unknown. You need to estimate means($\mu_-$: for negative class and  $\mu_+$: for positive class) and variances ($\sigma^{2}_{-}$: for negative class and $\sigma^{2}_+$: for positive class) from the given data : 
<br> (1) Assume $X|Y_{Y=0} \sim \mathcal{N}(\mu_- , \sigma^{2}_-)$ 
<br>(2) Assume $X|Y_{Y=1} \sim \mathcal{N}(\mu_+ , \sigma^{2}_+)$


*Generating artificial datasets in the next cell *

In [258]:
## This cell is for generating datasets. Students should not change anything in this cell. 
## You can compare your mean and variance estimates by the actual ones used to generate these datasets

import numpy as np
X_pos = np.random.randn(1000,1)+np.array([[2.]])
X_neg = np.random.randn(1000,1)+np.array([[4.]])
X_train_pos = X_pos[:900]
X_train_neg = X_neg[:900]
X_test_pos = X_pos[900:]
X_test_neg = X_neg[900:]
X_train = np.concatenate((X_train_pos, X_train_neg), axis=0)
X_test = np.concatenate((X_test_pos, X_test_neg), axis=0)
Y_train = np.concatenate(( np.ones(900),np.zeros(900) ))
Y_test = np.concatenate(( np.ones(100), np.zeros(100) ))

## X_train, X_test, Y_train, Y_test are your datasets to work with ####6



<br>**Instructions to follow for learning a Baeysian classifier:** *(Code the formulae for estimating the different parameters yourself)*
<br> a)Utilize the training dataset to estimate the means($\hat{\mu_+}$,$\hat{\mu_-}$) and variances($\hat{\sigma^{2}_+}$, $\hat{\sigma^{2}_-}$) for both positive and negative classes  
b)Estimate the prior probability: $P(Y=1)$  ⟶ which could be referred to as: $\hat{a}$ 
<br>c)Estimate the classifier funtion/posterior probability:  $P(Y=1|X = x)$  ⟶ which could be referred to as $\hat{\eta(x)}$
<br>d)Find out the threshold value($x^*$) for classification by equating the estimated classifier function($\hat{\eta(x)}$)  with threshold probability of 0.5
<br>e)Classify the test dataset into the two classes using this threshold value($x^*$) and find out the **accuracy** of the prediction 

Return back:  $\hat{\mu_+}$, $\hat{\mu_-}$, $\hat{\sigma^{2}_+}$, $\hat{\sigma^{2}_-}$, $\hat{a}$, $x^*$ and accuracy from the code written 

*Hint: $X|Y_{Y=0} \sim \mathcal{N}(\mu_- , \sigma^{2}_-)$ implies $P_{X|Y=0} = \mathcal{N}(\mu_- , \sigma^{2}_-) $*


### $\hat{\mu_+}$, $\hat{\mu_-}$, $\hat{\sigma^{2}_+}$, $\hat{\sigma^{2}_-}$ Calculation

In [263]:
def mean(b):
  sum=0
  length=0
  for i in range(int(len(b))):
    sum=sum+b[i]
    length=length+1
  mean=float(sum/length)
  return mean  
def variance(b):
  var = sum((i - mean(b)) ** 2 for i in b) / len(b)
  return var
mean_pos=mean(X_train_pos)
mean_neg=mean(X_train_neg)
var_pos=variance(X_train_pos)
var_neg=variance(X_train_neg)
print('μ+:' +str(mean_pos)+ ' μ−:' +str(mean_neg) + ' σ^2+:' + str(var_pos)+ ' σ^2−:'+str(var_neg))

μ+:2.050457042898035 μ−:4.016271974469526 σ^2+:[0.93375884] σ^2−:[0.96273269]


PRIOR PROBABILITY: $\hat{a}$

In [265]:
a=0
for i in range(len(Y_train)):
  if Y_train[i]==1:
    a=a+1
  else:
    continue 
P_1=a/len(Y_train)  ##Probability of P(Y=1)
P_0=1-P_1           ##Probability of P(Y=0)
print('Prior probability of y=1 is ', P_1)

Prior probability of y=1 is  0.5


POSTERIOR PROBABILITY

P(Y=1|X=x) = P(X=x,Y=1)*P(Y=1)/P(X)

P(X) = P(X=x,Y=1)+P(X=x,Y=0)

In [266]:
mu=mean_pos
std=np.sqrt(var_pos)
mu_0=mean_neg
std_0=np.sqrt(var_neg)
y_1=lambda x: ((np.exp(-np.power(x - mu, 2.) / (2 * np.power(std, 2.))))/(std*np.power(2*np.pi,0.5))) ##P(X=x,Y=1)
y_0=lambda x: ((np.exp(-np.power(x - mu_0, 2.) / (2 * np.power(std_0, 2.))))/(std_0*np.power(2*np.pi,0.5))) ##P(X=x,Y=0)
f=lambda x: (y_1(x)*0.5)/((y_1(x)+y_0(x))*0.5) ##posterior probability for Y=1; P(Y=1|X=x)

## The threshold value of $x^*$ for the probability to be 0.5. 
## The value is found to be **2.986**
---



In [ ]:
from scipy.optimize import minimize_scalar
import math
def func(x):
  return (f(x)-0.5)**2
res = minimize_scalar(func)
print(res.x)

[2.98690673]


array([0.49342184])

## Predictions:

Anything above x=2.986 will have posterior probability for y=1 less than 0.5, which means y=0. Hence any x greater than x=2.986 will be predicted as y=0 and any x less than 2.986 will be predicted as y=1.

In [ ]:
y_pred=[]
for i in range(len(X_test)):
  if X_test[i]>=2.986:
    y_pred.append(0)
  else:
    y_pred.append(1)

ACCURACY

In [ ]:
count=0
for i in range(len(y_pred)):
  if y_pred[i]==Y_test[i]:
    count=count+1
  else:
    continue
print('The accuracy of the prediction is ',count/len(Y_test))

The accuracy of the prediction is  0.9


### From Discrete distribution of data

Unlike the first exercise for learning the Naive Bayes' classifier where we dealt with continuous distribution of data, here you need to work with discrete data, which means finding Probability Mass Distribution(PMF). 

Age  | Income | Status  | Buy
-----|--------|-------- |----
<=20 |  low   | students| yes
<=20 |  high  | students| yes
<=20 | medium | students| no
<=20 | medium | married | no
<=20 |  high  | married | yes
21-30|  low   | married | yes
21-30|  low   | married | no 
21-30| medium | students| no
21-30|  high  | students| yes
 >30 |  high  | married | no
 >30 |  high  | married | yes
 >30 | medium | married | yes
 >30 | medium | married | no
 >30 | medium | students| no
 
Consider the train dataset above. Take any random datapoint ($X_{i}$) where $X_{i} = (X_{i,1} = Age,X_{i,2} = Income,X_{i,3} = Status)$ and its corresponding label 

($Y_{i} = Buy$). A "yes" in Buy corresponds to label-1 and a "no" in Buy corresponds to label-0.

<br>**Instructions to follow for learning a Baeysian classifier:** *(Code the formulae for estimating the different parameters yourself)*
<br> a)Estimate the prior probability: $P(Y=1)$  ⟶ which could be referred to as: $\hat{a}$   
b)Estimate the likelihood for each feature:  $P(X_{i,j} = x |Y = y_{i})$, where $ i$=datapoint counter, $j \in \{1,2,3\}$ and $y_{i} \in \{0,1\}$ 
<br>c)Estimate the total likelihood: $P(X_{i} = x |Y = y_{i})$  
d)Calculate the posterior probability: $P(Y = 1|X_{i} = x_{test} )$ = $p_{test}$ where $x_{test} = (Age = 21-30, Income= medium, Status = married)$


Return back: $\hat{a}$, total likelihood and $p_{test}$ 


In [243]:
d={'Age':['<=20','<=20','<=20','<=20','<=20','21-30','21-30','21-30','21-30','>30','>30','>30','>30','>30'],'Income':['low','high','medium','medium','high','low','low','medium','high','high','high','medium','medium','medium'],'Status':['students','students','students','married','married','married','married','students','students','married','married','married','married','students'],'Buy':['yes','yes','no','no','yes','yes','no','no','yes','no','yes','yes','no','no']}
data=pd.DataFrame(d)
data

Age  Income    Status  Buy
0    <=20     low  students  yes
1    <=20    high  students  yes
2    <=20  medium  students   no
3    <=20  medium   married   no
4    <=20    high   married  yes
5   21-30     low   married  yes
6   21-30     low   married   no
7   21-30  medium  students   no
8   21-30    high  students  yes
9     >30    high   married   no
10    >30    high   married  yes
11    >30  medium   married  yes
12    >30  medium   married   no
13    >30  medium  students   no

## prior probability: $P(Y=1)$: $\hat{a}$ = 0.5


In [210]:
data[data['Buy']=='yes']

Age  Income    Status  Buy
0    <=20     low  students  yes
1    <=20    high  students  yes
4    <=20    high   married  yes
5   21-30     low   married  yes
8   21-30    high  students  yes
10    >30    high   married  yes
11    >30  medium   married  yes

In [211]:
P_y_1=len(data[data['Buy']=='yes'])/len(data)

##likelihood for each feature:  $P(X_{i,j} = x |Y = y_{i})$

### AGE: ['<=20' '21-30' '>30'] for Y=Yes

In [212]:

a=data[['Age','Buy']]
a=data[data['Buy']=='yes']
likelihood_age_yes={}
for i in a['Age'].unique():
  likelihood_age_yes[i]=len(a[a['Age']==i])/len(a)
likelihood_age_yes

{'21-30': 0.2857142857142857,
 '<=20': 0.42857142857142855,
 '>30': 0.2857142857142857}

### AGE: ['<=20' '21-30' '>30'] for Y=No

In [213]:
a=data[['Age','Buy']]
a=data[data['Buy']=='no']
likelihood_age_no={}
for i in a['Age'].unique():
  likelihood_age_no[i]=(len(a[a['Age']==i])/len(a))
likelihood_age_no

{'21-30': 0.2857142857142857,
 '<=20': 0.2857142857142857,
 '>30': 0.42857142857142855}

## INCOME ['low' 'high' 'medium'] for Y=yes

In [214]:
a=data[['Income','Buy']]
a=data[data['Buy']=='yes']
likelihood_income_yes={}
b=a['Income'].unique()
for i in b:
  likelihood_income_yes[i]=(len(a[a['Income']==i])/len(a))
likelihood_income_yes

{'high': 0.5714285714285714,
 'low': 0.2857142857142857,
 'medium': 0.14285714285714285}

## INCOME ['low' 'high' 'medium'] for Y=no

In [215]:
a=data[['Income','Buy']]
a=data[data['Buy']=='no']
likelihood_income_no={}
for i in b:
  likelihood_income_no[i]=(len(a[a['Income']==i])/len(a))
likelihood_income_no


{'high': 0.14285714285714285,
 'low': 0.14285714285714285,
 'medium': 0.7142857142857143}

## STATUS ['students' 'married'] for Y=yes

In [216]:
a=data[['Status','Buy']]
a=data[data['Buy']=='yes']
likelihood_status_yes={}
b=a['Status'].unique()
for i in b:
  likelihood_status_yes[i]=(len(a[a['Status']==i])/len(a))
likelihood_status_yes

{'married': 0.5714285714285714, 'students': 0.42857142857142855}

## STATUS ['students' 'married'] for Y=no

In [217]:
a=data[['Status','Buy']]
a=data[data['Buy']=='no']
likelihood_status_no={}
b=a['Status'].unique()
for i in b:
  likelihood_status_no[i]=(len(a[a['Status']==i])/len(a))
likelihood_status_no

{'married': 0.5714285714285714, 'students': 0.42857142857142855}

## TOTAL LIKELIHOOD:


Ex:$ P([<=20,low,married],Y=yes) =$ $P(<=20,Y=yes)*P(low,Y=yes)*P(married,Y=yes)$

Similarly we can calculate for all possibilities:
A few are listed below

In [218]:
likelihood_age_yes['<=20']*likelihood_income_yes['low']*likelihood_status_yes['married']

0.06997084548104955

1) P(<=20,low,students,Y=yes)

In [219]:
likelihood_age_yes['<=20']*likelihood_income_yes['low']*likelihood_status_yes['students']

0.05247813411078716

1) P(<=20,medium,married,Y=yes)

In [220]:
likelihood_age_yes['<=20']*likelihood_income_yes['medium']*likelihood_status_yes['married']

0.034985422740524776

1) P(<=20,medium,students,Y=yes)

In [221]:
likelihood_age_yes['<=20']*likelihood_income_yes['medium']*likelihood_status_yes['students']

0.02623906705539358

1) P(<=20,high,married,Y=yes)

In [222]:
likelihood_age_yes['<=20']*likelihood_income_yes['high']*likelihood_status_yes['married']

0.1399416909620991

1) P(<=20,high,students,Y=yes)

In [223]:
likelihood_age_yes['<=20']*likelihood_income_yes['high']*likelihood_status_yes['students']

0.10495626822157432

## Total Likelihood for xtest is 

## P(21-30,medium,married,Y=yes)

In [269]:
total_likelihood=likelihood_age_yes['21-30']*likelihood_income_yes['medium']*likelihood_status_yes['married']

#Prediction-Posterior Probability
$P(Y = 1|X_{i} = x_{test} )$ = $p_{test}$ where $x_{test} = (Age = 21-30, Income= medium, Status = married)$

$P(Y = 1|X_{i} = x_{test} )$ = $P(X_{i} = x_{test} | Y=1 ) * P(Y=1) / P(X_{i} = X_{test})$

$P(X_{i} = X_{test})$ =  $P(X_{i} = x_{test} | Y=1 )*P(Y=1) + P(X_{i} = x_{test} | Y=0 )*P(Y=0)$ 



In [272]:
ptest=(total_likelihood*P_y_1)/(likelihood_age_yes['21-30']*likelihood_income_yes['medium']*likelihood_status_yes['married']*P_y_1+likelihood_age_no['21-30']*likelihood_income_no['medium']*likelihood_status_no['married']*(1-P_y_1))

In [273]:
ptest

0.16666666666666663

#### The probability for Y=1 is 0.166666 which is less than 0.5, hence the output is predicted as Y=0
 